### DISCUSSION TITLE
Bit Depth switching support

### DISCUSSION THREAD

**Discussion by vincentneo** [vincentneo/LosslessSwitcher#62] [Announcements]
Happy New Year!
Experimental build that switches both sample rate (e.g. 192kHz) and **bit depth** (e.g. 16-bit/24-bit), is now here.

Those facing issue #2, please look here!

## Known Issues
- Menu does not show bit depth info yet, please use Audio MIDI Setup to observe if bit depth has switched.
- Detection may not be as good as prior builds ~~(has yet to verify impact of detection, just a guess)~~, as I've disabled a detection method via `com.apple.coremedia` messages due to lack of bit depth info. 
- Use of `com.apple.coreaudio` messages now, known to not provide info for certain sample rates in the past, but have yet to observe that in Ventura 13.1 so far.
- Code is a mess right now! 

<hr /><em>This discussion was created from the release <a href='https://github.com/vincentneo/LosslessSwitcher/releases/tag/1.1.0-beta8'>Bit Depth switching support</a>.</em>

**Comment by MasterRahool** [comment]
"Detection may not be as good as prior builds"

Sums things up. Spotty detective of sample rate now, often click on a hi-res 96k track and LSS selects 48k instead. Depth seems to be switching, when things switch, though. I kept local track detection off because that just added more problems.

**Reply by vincentneo** [reply]
Seems like right now it's a question of whether if you want bit depth switching or better detection. Don't think there's an in between option at all.

Maybe I will let the user decide for themselves, like the local track detection support.

**Reply by MasterRahool** [reply]
> Seems like right now it's a question of whether if you want bit depth switching or better detection. Don't think there's an in between option at all.
> 
> Maybe I will let the user decide for themselves, like the local track detection support.

does local track detection ever work right? i do have a small amount of local ALAC rips but it defeats the purpose if i have to always check it when using local tracks and uncheck it when playing the next album, or using it in a mixed playlist. So to date I just leave it off.

*personally* i dont want or need bit depth changing, but it would be cool to have the track's depth in the menu bar icon thing beside the rate

**Reply by vincentneo** [reply]
> personally i dont want or need bit depth changing, but it would be cool to have the track's depth in the menu bar icon thing beside the rate

Eventually I think I will make it a setting.

Like you could set if you just want the UI to show this or detection to run this, etc.

**Reply by neo2049** [reply]
Better detection would be preferable to bit depth.

**Comment by Floschi12345** [comment]
Please make Bitrate Switching optional. For those of us who use ONLY digital volume control, i.e. with headphone dongles etc. or DSPs the extra bitrate padding is essential not to lose dynamics/overhead. 8bits of padding are equal to a -48db reduction in digital volume control, so if we playback a 16bit file at 24bit we can reduce the volume digitally by -48db without losing any dynamics or truncating the actual audio content. Same goes for playing 24bit content at 32bit. So bitrate switching makes sense when outputting a fullrange signal to a device with analogue volume control, otherwise not so much.

**Reply by vincentneo** [reply]
If I understand correctly, you mean an option to remain as it was, in previous builds / v1.0, am I right?

**Reply by Floschi12345** [reply]
Yes, make an option to make bitdepth switching optional, as there are (as mentioned above) are scenarios where a bitdepth padding is very useful. Sorry, I was typing very early in the morning. You have to exchange bitrate for bitdepth in my first post...;-)

**Reply by vincentneo** [reply]
@Floschi12345 thanks for clarifying!

**Comment by DeckHeadFi** [comment]
Thanks for adding this feature of switching bitdepth!! This was the missing feature. I feel like I'm getting the most out of Apple Music finally. Loving it!!

The only thing I think truly that could be smoother so far is that you're stopping the song to change the sample rate and bit depth is great in theory. It's just on the jarring side because the song will play for a very quick moment, then stop, and change the sample rate/bit depth. Which is still much better than not stopping the song. It would just be ideal of course if you could have a pause before the song plays at all. And perhaps a longer pause too for a couple seconds or so to make sure everything is good to go.

**Reply by vincentneo** [reply]
>  It's just on the jarring side because the song will play for a very quick moment, then stop, and change the sample rate/bit depth.

Well, LS works by looking into logs from Apple Music app. The log will only start showing messages that contains the sample rate info once the song starts playing. So essentially, there is no way to implement it as ideally as possible right now, unless there's other ways to detect the info, so sorry about that.

**Reply by DeckHeadFi** [reply]
You may have already looked into this, but instead of waiting for the sample rate info to come in (which can take a second or longer into the song), perhaps stop every song for 1-2 seconds right at 0:00 to give the program more time to read the sample rate/bit depth. I don't know if it's possible to read the info at the 0:00 mark, or if you can mute the volume on your Mac during that split second if you need it. What is currently happening (seems like), you're waiting for the info to come in then stopping the song which is why it's different for different songs as it's probably not always perfectly quick. So yeah either stop every song at 0:00 and give yourself time to read the info or stop it at 0:01 while muting the start of the song. These could all be toggles for the user, but just an idea to be less jarring.

**Reply by vincentneo** [reply]
> You may have already looked into this

Actually, I haven't looked into that. Thanks for the idea!

**Reply by troylefevre65** [reply]
All Apologies if this has already been addressed. I’m new to the project.  Firstly, my gratitude for  your development efforts!  I’m fully aware how such can be a real time-sink. Offering development time, for only voluntarily monetary reward is greatly appreciated. 

From my limited experience with the product, I’m somewhat confused by the above poster’s assertion, and the author’s validation of, the delayed knowledge of bitrate...  Not infrequently, my DAC (Schiit Yggdrasil OG) will switch bitrates mid-stream during the currently playing song.  When this happens, the bitrate then reported by Lossless Switcher is errant to that of the currently playing song reported by Apple Music.  However, when the stream advanced to the very next song, what was just previously erroneous, is now accurate, where LS and Apple Music align?  

This behavior leads me to believe that there is knowledge of the next song’s bitrate, ahead of the next song’s beginning?

My usual use case is a Siri request like, “play music I like” to start a listening session.  Where I’m a pure Apple Music subscriber with no local tracks.

**Reply by vincentneo** [reply]
@troylefevre65 LosslessSwitcher works by reading off logs written by the Apple Music app, every few seconds (for v1.0) and for the first few seconds only (for betas, mostly). 
If there was something new detected from the logs, LosslessSwitcher will switch the sample rate of the selected output device.

LS gets confused however, when the Apple Music app suddenly emits logs for the next track, at the wrong timing. Due to the nature of the logs, there's no way to guarantee which log message refers to which track.

I hope that clears things up a little!

**Reply by troylefevre65** [reply]
@vincentneo  Thank you for your response.  That does clear it up.  I suspected that might be the case.  What other metadata do the Apple Music logs inform?

**Reply by ianbrianmorgan** [reply]
Hello and thanks in advance for working on this project @vincentneo!

I think I'm experiencing a similar thing as @troylefevre65. I'v noticed that multiple songs will down sample in the middle of the song. The one I've found to be pretty consistent is "The Way You Make Me Feel (Single Version) from Michael Jackson.

Computer is Apple MacBook Pro with M1 Max (2021 model). 

Apple Music setup is Dolby Atmos set to off. Sound Check is on Lossless set for ALAC. While I've had this problem since using the app with my REM ADI-2 FS DAC, this time I used my Topping stack to compare. A couple songs played fine, then on the 3rd song ("The Way You Make Me Feel (Single Version) from Michael Jackson, it dropped sampling from 96kHz to 44.1kHz at the 3 minute mark. I then plugged my Topping stack directly into my MacBook (I was using CalDigit Thunderbolt 3 hub). Same thing happened. Then I downloaded the song and then played it again--no issues. Deleted the download and played it again--dropped to 44.1kHz at the 3 min mark again. This seems to be consistent with streaming using the MacOS Apple Music app. MacBook is on LAN connection and I highly doubt an issue with the bandwidth/quality because I stream perfect from Tidal (including MQA). I then also tried from my iPhone (on same WiFI network) using lossless/ALAC to my Topping via the Apple Camera Connection Kit and this same song started out fine at 96kHz. Song played and stayed at 96kHz throughout. Obviously, LosslessSwitcher is not on my iOS device ;-)

So I'm thinking this has something with the LosslessSwitcher app that automatically changes the output sample rate of the DAC audio in Audio MIDI Setup to match the source sample rate. I killed off the LoslessSwitcher app and manually changed the sampling rate to see if I still saw the same downsample. Never happened.  

I'd also echo the poster's suggestion to spool up the song for a couple seconds to allow LosslessSwitcher to change the sample rate BEFORE audio starts playing. Currently, the audio starts for as second or two, then cuts out if the sample rate needs to be changed. If this was done like Tidal and DAC exclusive mode, there would not be detectable audio pause.

Lastly, a version number on the app would be awesome. I was not able to find that so I could let you know the version I'm using. I downloaded in for the first time a few weeks ago, so it's fairly recent. 

Anyway--thanks a million for you spending time on this MASSIVE oversight by Apple.
